File: src/main/java/ro/mpp2024/repository/ReservedSeatRepository.java
package ro.mpp2024.repository;

import ro.mpp2024.model.Employee;
import ro.mpp2024.model.ReservedSeat;
import ro.mpp2024.model.Trip;

import java.sql.*;
import java.util.*;

public class ReservedSeatRepository extends AbstractRepository<Integer, ReservedSeat> implements IReservedSeatRepository {

    private final TripRepository tripRepository;
    private final EmployeeRepository employeeRepository;
    private final ClientRepository clientRepository;

    public ReservedSeatRepository(Properties props,
                                  TripRepository tripRepository,
                                  EmployeeRepository employeeRepository,
                                  ClientRepository clientRepository) {
        super(props);
        this.tripRepository = tripRepository;
        this.employeeRepository = employeeRepository;
        this.clientRepository = clientRepository;
    }

    @Override
    public Optional<ReservedSeat> findById(Integer id) {
        logger.info("Find ReservedSeat by ID: {}", id);
        String query = "SELECT * FROM ReservedSeats WHERE id = ?";
        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setInt(1, id);
            ResultSet resultSet = statement.executeQuery();

            if (resultSet.next()) {
                return extractReservedSeatFromResultSet(resultSet);
            }
        } catch (SQLException e) {
            logger.error("Database error while finding ReservedSeat with id {}", id, e);
        }
        return Optional.empty();
    }

    @Override
    public List<ReservedSeat> findAll() {
        logger.info("Find all ReservedSeats");
        List<ReservedSeat> reservedSeats = new ArrayList<>();
        String query = "SELECT * FROM ReservedSeats";
        try (Connection connection = jdbc.getConnection();
             Statement statement = connection.createStatement();
             ResultSet resultSet = statement.executeQuery(query)) {

            while (resultSet.next()) {
                extractReservedSeatFromResultSet(resultSet).ifPresent(reservedSeats::add);
            }
        } catch (SQLException e) {
            logger.error("Database error while finding all ReservedSeats", e);
        }
        return reservedSeats;
    }

    public List<ReservedSeat> findByTripId(Integer tripId) {
        logger.info("Find ReservedSeats by Trip ID: {}", tripId);
        List<ReservedSeat> reservedSeats = new ArrayList<>();
        String query = "SELECT * FROM ReservedSeats WHERE trip_id = ?";
        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setInt(1, tripId);
            ResultSet resultSet = statement.executeQuery();

            while (resultSet.next()) {
                extractReservedSeatFromResultSet(resultSet).ifPresent(reservedSeats::add);
            }
        } catch (SQLException e) {
            logger.error("Database error while finding ReservedSeats for trip id {}", tripId, e);
        }
        return reservedSeats;
    }

    @Override
    public Optional<ReservedSeat> save(ReservedSeat reservedSeat) {
        String query = "INSERT INTO ReservedSeats(trip_id, employee_id, seat_number, client_id) VALUES(?, ?, ?, ?)";
        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS)) {

            statement.setInt(1, reservedSeat.getTrip().getId());
            statement.setInt(2, reservedSeat.getEmployee().getId());
            statement.setInt(3, reservedSeat.getSeatNumber());
            statement.setInt(4, reservedSeat.getClient().getId());
            statement.executeUpdate();

            ResultSet resultSet = statement.getGeneratedKeys();
            if (resultSet.next()) {
                reservedSeat.setId(resultSet.getInt(1));
                return Optional.of(reservedSeat);
            }
        } catch (SQLException e) {
            logger.error("Database error while saving ReservedSeat", e);
        }
        return Optional.empty();
    }

    @Override
    public Optional<ReservedSeat> delete(Integer id) {
        logger.info("Delete ReservedSeat with ID: {}", id);
        Optional<ReservedSeat> reservedSeatToDelete = findById(id);
        if (reservedSeatToDelete.isEmpty()) {
            return Optional.empty();
        }

        String query = "DELETE FROM ReservedSeats WHERE id = ?";
        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setInt(1, id);
            int affectedRows = statement.executeUpdate();
            if (affectedRows > 0) {
                return reservedSeatToDelete;
            }
        } catch (SQLException e) {
            logger.error("Database error while deleting ReservedSeat with id {}", id, e);
        }
        return Optional.empty();
    }

    @Override
    public Optional<ReservedSeat> update(ReservedSeat reservedSeat) {
        String query = "UPDATE ReservedSeats SET trip_id=?, employee_id=?, seat_number=?, client_id=? WHERE id=?";
        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setInt(1, reservedSeat.getTrip().getId());
            statement.setInt(2, reservedSeat.getEmployee().getId());
            statement.setInt(3, reservedSeat.getSeatNumber());
            statement.setInt(4, reservedSeat.getClient().getId());
            statement.setInt(5, reservedSeat.getId());

            int affectedRows = statement.executeUpdate();
            if (affectedRows > 0) {
                return Optional.of(reservedSeat);
            }
        } catch (SQLException e) {
            logger.error("Database error while updating ReservedSeat with id {}", reservedSeat.getId(), e);
        }
        return Optional.empty();
    }

    private Optional<ReservedSeat> extractReservedSeatFromResultSet(ResultSet resultSet) throws SQLException {
        Integer id = resultSet.getInt("id");
        Integer tripId = resultSet.getInt("trip_id");
        int employeeId = resultSet.getInt("employee_id");
        Integer seatNumber = resultSet.getInt("seat_number");
        int clientId = resultSet.getInt("client_id");

        Optional<Trip> tripOpt = tripRepository.findById(tripId);
        if (tripOpt.isPresent()) {
            ReservedSeat reservedSeat = new ReservedSeat();
            reservedSeat.setId(id);
            reservedSeat.setTrip(tripOpt.get());
            reservedSeat.setSeatNumber(seatNumber);

            if (employeeId != 0 && !resultSet.wasNull()) {
                employeeRepository.findById(employeeId).ifPresent(reservedSeat::setEmployee);
            }

            if (clientId != 0 && !resultSet.wasNull()) {
                clientRepository.findById(clientId).ifPresent(reservedSeat::setClient);
            }

            return Optional.of(reservedSeat);
        }
        return Optional.empty();
    }

    @Override
    public List<ReservedSeat> findByTripDestinationDateTime(String destination, String date, String time) {
        logger.info("Find ReservedSeat by trip destination, date and time: {}: {}, {}", destination, date, time);
        List<ReservedSeat> reservedSeats = new ArrayList<>();
        String query = "SELECT * FROM ReservedSeats rs " +
                "JOIN Trip t ON rs.trip_id = t.id " +
                "JOIN Destination d ON t.destination_id = d.id " +
                "WHERE d.name = ? AND t.departure_date = ? AND t.departure_time = ?";
        try (Connection connection = jdbc.getConnection();
            PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setString(1, destination);
            System.out.println(date);
            System.out.println(time);
            statement.setString(2, date);
            statement.setString(3, time);

            ResultSet resultSet = statement.executeQuery();

            while (resultSet.next()) {
                extractReservedSeatFromResultSet(resultSet).ifPresent(reservedSeats::add);
            }
        } catch (SQLException e) {
            logger.error("Database error while finding ReservedSeat by trip destination, date and time: {}: {}, {}", destination, date, time, e);
        }
        return reservedSeats;
    }

//    @Override
//    public boolean reserveSeats(String clientName, List<Integer> seatNumbers, Trip trip, Employee employee) {
//        logger.info("Reserve seats for client {} on trip {}", ., clientName, trip);
//        String query = "INSERT INTO ReservedSeats(trip_id, employee_id, seat_number, client_id, reservation_date) VALUES(?, ?, ?, ?, ?)";
//    }
}

File: csharpServer/persistance/ReservedSeatRepository.cs
using System;
using System.Collections.Generic;
using Microsoft.Data.Sqlite;
using Microsoft.Extensions.Logging;
using model;

namespace persistance
{
    public class ReservedSeatRepository : AbstractRepository<int, ReservedSeat>, IReservedSeatRepository
    {
        private readonly ITripRepository tripRepository;
        private readonly IEmployeeRepository employeeRepository;
        private readonly IClientRepository clientRepository;

        public ReservedSeatRepository(ITripRepository tripRepository, IEmployeeRepository employeeRepository, IClientRepository clientRepository) : base()
        {
            this.tripRepository = tripRepository;
            this.employeeRepository = employeeRepository;
            this.clientRepository = clientRepository;
        }

        private ReservedSeat? ExtractReservedSeatFromResultSet(SqliteDataReader reader)
        {
            try
            {
                var id = reader.GetInt32(reader.GetOrdinal("id"));
                var tripId = reader.GetInt32(reader.GetOrdinal("trip_id"));
                var employeeId = reader.GetInt32(reader.GetOrdinal("employee_id"));
                var seatNumber = reader.GetInt32(reader.GetOrdinal("seat_number"));
                var clientId = reader.GetInt32(reader.GetOrdinal("client_id"));

                var trip = tripRepository.FindById(tripId)
                    ?? throw new InvalidOperationException($"Trip with ID {tripId} not found");

                var reservedSeat = new ReservedSeat
                {
                    Id = id,
                    Trip = trip,
                    SeatNumber = seatNumber
                };

                if (employeeId != 0)
                {
                    var emp = employeeRepository.FindById(employeeId);
                    if (emp != null)
                    {
                        reservedSeat.Employee = emp;
                    }
                }
                if (clientId != 0)
                {
                    var client = clientRepository.FindById(clientId);
                    if (client != null)
                    {
                        reservedSeat.Client = client;
                    }
                }
                Console.WriteLine(reservedSeat);
                return reservedSeat;
            }
            catch (SqliteException e)
            {
                logger.LogError(e, "Error while extracting ReservedSeat from ResultSet");
                return null;
            }
        }

        public override ReservedSeat? FindById(int id)
        {
            logger.LogInformation("Find ReservedSeat by ID: {Id}", id);
            const string query = "SELECT * FROM ReservedSeats WHERE id = @id";
            try
            {
                using (var connection = jdbc.GetConnection())
                using (var command = new SqliteCommand(query, (SqliteConnection)connection))
                {
                    command.Parameters.AddWithValue("@id", id);
                    connection.Open();
                    using (var reader = command.ExecuteReader())
                    {
                        if (reader.Read())
                        {
                            return ExtractReservedSeatFromResultSet(reader);
                        }
                    }
                }
            }
            catch (SqliteException e)
            {
                logger.LogError(e, "Database error while finding ReservedSeat with ID {Id}", id);
            }
            return null;
        }

        public override IEnumerable<ReservedSeat> FindAll()
        {
            logger.LogInformation("Find all ReservedSeats");
            var reservedSeats = new List<ReservedSeat>();
            const string query = "SELECT * FROM ReservedSeats";
            try
            {
                using (var connection = jdbc.GetConnection())
                using (var command = new SqliteCommand(query, (SqliteConnection)connection))
                {
                    connection.Open();
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            var reservedSeat = ExtractReservedSeatFromResultSet(reader);
                            if (reservedSeat != null)
                            {
                                reservedSeats.Add(reservedSeat);
                            }
                        }
                    }
                }
            }
            catch (SqliteException e)
            {
                logger.LogError(e, "Database error while finding all ReservedSeats");
            }
            return reservedSeats;
        }

        public IEnumerable<ReservedSeat> FindByTripId(int tripId)
        {
            logger.LogInformation("Find ReservedSeats by Trip ID: {TripId}", tripId);
            var reservedSeats = new List<ReservedSeat>();
            const string query = "SELECT * FROM ReservedSeats WHERE trip_id = @trip_id";
            try
            {
                using (var connection = jdbc.GetConnection())
                using (var command = new SqliteCommand(query, (SqliteConnection)connection))
                {
                    command.Parameters.AddWithValue("@trip_id", tripId);
                    connection.Open();
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            var reservedSeat = ExtractReservedSeatFromResultSet(reader);
                            if (reservedSeat != null)
                            {
                                reservedSeats.Add(reservedSeat);
                            }
                        }
                    }
                }
            }
            catch (SqliteException e)
            {
                logger.LogError(e, "Database error while finding ReservedSeats for trip ID {TripId}", tripId);
            }
            return reservedSeats;
        }

        public override ReservedSeat? Save(ReservedSeat reservedSeat)
        {
            logger.LogInformation("Save ReservedSeat: {ReservedSeat}", reservedSeat);
            const string query = "INSERT INTO ReservedSeats (trip_id, employee_id, seat_number, client_id) VALUES (@trip_id, @employee_id, @seat_number, @client_id)";
            try
            {
                using (var connection = jdbc.GetConnection())
                {
                    connection.Open();
                    using (var command = new SqliteCommand(query, (SqliteConnection)connection))
                    {
                        command.Parameters.AddWithValue("@trip_id", reservedSeat.Trip.Id);
                        command.Parameters.AddWithValue("@employee_id", reservedSeat.Employee.Id);
                        command.Parameters.AddWithValue("@seat_number", reservedSeat.SeatNumber);
                        command.Parameters.AddWithValue("@client_id", reservedSeat.Client.Id);
                        command.ExecuteNonQuery();
                    }
                    using (var idCommand = new SqliteCommand("SELECT last_insert_rowid()", (SqliteConnection)connection))
                    {
                        var id = Convert.ToInt32(idCommand.ExecuteScalar());
                        reservedSeat.Id = id;
                        return reservedSeat;
                    }
                }
            }
            catch (SqliteException e)
            {
                logger.LogError(e, "Database error while saving ReservedSeat: {ReservedSeat}", reservedSeat);
            }
            return null;
        }

        public override ReservedSeat? Delete(int id)
        {
            logger.LogInformation("Delete ReservedSeat with ID: {Id}", id);
            var reservedSeatToDelete = FindById(id);
            if (reservedSeatToDelete == null)
            {
                return null;
            }
            const string query = "DELETE FROM ReservedSeats WHERE id = @id";
            try
            {
                using (var connection = jdbc.GetConnection())
                using (var command = new SqliteCommand(query, (SqliteConnection)connection))
                {
                    command.Parameters.AddWithValue("@id", id);
                    connection.Open();
                    var affectedRows = command.ExecuteNonQuery();
                    if (affectedRows > 0)
                    {
                        return reservedSeatToDelete;
                    }
                }
            }
            catch (SqliteException e)
            {
                logger.LogError(e, "Database error while deleting ReservedSeat with ID {Id}", id);
            }
            return null;
        }

        public override ReservedSeat? Update(ReservedSeat reservedSeat)
        {
            logger.LogInformation("Update ReservedSeat: {ReservedSeat}", reservedSeat);
            const string query = "UPDATE ReservedSeats SET trip_id = @trip_id, employee_id = @employee_id, seat_number = @seat_number, client_id = @client_id WHERE id = @id";
            try
            {
                using (var connection = jdbc.GetConnection())
                using (var command = new SqliteCommand(query, (SqliteConnection)connection))
                {
                    command.Parameters.AddWithValue("@trip_id", reservedSeat.Trip.Id);
                    command.Parameters.AddWithValue("@employee_id", reservedSeat.Employee.Id);
                    command.Parameters.AddWithValue("@seat_number", reservedSeat.SeatNumber);
                    command.Parameters.AddWithValue("@client_id", reservedSeat.Client.Id);
                    command.Parameters.AddWithValue("@id", reservedSeat.Id);
                    connection.Open();
                    var affectedRows = command.ExecuteNonQuery();
                    if (affectedRows > 0)
                    {
                        return reservedSeat;
                    }
                }
            }
            catch (SqliteException e)
            {
                logger.LogError(e, "Database error while updating ReservedSeat: {ReservedSeat}", reservedSeat);
            }
            return null;
        }

        public List<ReservedSeat> FindByTripDestinationDateTime(string destination, string date, string time)
        {
            logger.LogInformation("Find ReservedSeat by trip destination, date and time: {Destination}: {Date}, {Time}", destination, date, time);
            var reservedSeats = new List<ReservedSeat>();
            const string query = @"
        SELECT * FROM ReservedSeats rs
        JOIN Trip t ON rs.trip_id = t.id
        JOIN Destination d ON t.destination_id = d.id
        WHERE d.name = @destination AND t.departure_date = @date AND t.departure_time = @time";
            try
            {
                using (var connection = jdbc.GetConnection())
                using (var command = new SqliteCommand(query, (SqliteConnection)connection))
                {
                    command.Parameters.AddWithValue("@destination", destination);
                    command.Parameters.AddWithValue("@date", date);
                    command.Parameters.AddWithValue("@time", time);
                    connection.Open();
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            var reservedSeat = ExtractReservedSeatFromResultSet(reader);
                            if (reservedSeat != null)
                            {
                                reservedSeats.Add(reservedSeat);
                            }
                        }
                    }
                }
            }
            catch (SqliteException e)
            {
                logger.LogError(e, "Database error while finding ReservedSeat by trip destination, date and time: {Destination}: {Date}, {Time}", destination, date, time);
            }
            return reservedSeats;
        }
    }
}